src/notebook/notebookFeature.ts

```diff
@@ -49,11 +69,39 @@ export class JuliaNotebookFeature {
                 }
             })
             controller.executeHandler = this.executeCells.bind(this)
+            controller.onDidChangeSelectedNotebooks(({ notebook, selected }) => {
+                // If we select our controller, then update the notebook metadata with the kernel information.
+                if (!selected) {
+                    return;
+                }
+                this.updateNotebookWithSelectedKernel(notebook, displayName, ver);
+            }, this, this.disposables)
 
-            this._controllers.push(controller)
+            this._controllers.set(controller, {version: ver })
         }
     }
 
+    private onDidOpenNotebookDocument(e: vscode.NotebookDocument) {
+        if (!this.isJuliaNotebook(e) || this._controllers.size === 0) {
+            return;
+        }
+        // Get metadata from notebook (to get an hint of what version of julia is used)
+        const { name, version } = this.getKernelSpecNameAndVersion(e);
+        let preferredControllerFound = false;
+        this._controllers.forEach((info, controller) => {
+            // If we find a controller that matches the vesion in the notebook metadata, then set
+            // that controller as the preferred controller.
+            if (name.includes(info.version) || version.includes(info.version)) {
+                preferredControllerFound = true;
+                controller.updateNotebookAffinity(e, vscode.NotebookControllerAffinity.Preferred)
```

Comment:
Magic sauce, this is waht puts Julia on the top of the list.
To test, please open an existing Jualia notebook (created in Jupyter), & ensure you delete (temporarily move) the ijulia kernelspec .json file.

Or you can wait for the CI to complete & download the Jupyter VSIX from here https://github.com/microsoft/vscode-jupyter/actions/runs/1006379241 & test it all together.